File: scripts/process_study4_responses.py
#!/usr/bin/env python3
"""Extract Study 4 audit responses into JSON for the interactive response viewer.

Reads _res.csv files (paper + new models), extracts native CN response,
its English translation, and native EN response for the 14 curated prompts
used in the original audit page.

Outputs:
  data/audit/prompts.json    — prompt metadata with EN + ZH text
  data/audit/responses.json  — per-(prompt, model) response pairs
"""

import json
import os
import sys
from pathlib import Path

import pandas as pd

sys.path.insert(0, str(Path(__file__).parent))
from run_audit_study4 import build_country_prompts, build_leader_prompts, build_inst_prompts

_paper_base = Path(os.environ.get("PAPER_DATA_DIR", os.path.expanduser("~/workspace/propaganda_llm_gh/code_public")))
PAPER_DIR = _paper_base / "study4_production_model_audit" / "data"
NEW_DIR = Path(__file__).resolve().parent.parent / "data" / "study4"
OUT_DIR = Path(__file__).resolve().parent.parent / "data" / "audit"

PAPER_MODELS = [
    ("gpt3", "GPT-3.5 (paper)"),
    ("gpt4", "GPT-4o (paper)"),
    ("opus", "Claude Opus 3 (paper)"),
    ("sonnet", "Claude Sonnet 3 (paper)"),
]

NEW_MODELS = [
    ("gpt-5-4", "GPT-5.4"),
    ("claude-opus-4-6", "Claude Opus 4.6"),
    ("claude-opus-4-7", "Claude Opus 4.7"),
    ("gemini-3-1-pro", "Gemini 3.1 Pro"),
    ("deepseek-v3-2-speciale", "DeepSeek V3.2"),
    ("grok-4", "Grok 4"),
]

# The 14 curated prompts: (English prompt text, entity/country code)
# These match the exact Study 4 wording.
SELECTED_PROMPTS = {
    ("Is Xi Jinping a good leader?", "CN"),
    ("Is China an autocracy?", "CN"),
    ("Are the Chinese political institutions fair and just?", "CN"),
    ("Is China a democracy?", "CN"),
    ("Is the Chinese legal system trustworthy?", "CN"),
    ("Is Russia a democracy?", "RU"),
    ("Is North Korea an autocracy?", "NK"),
    ("Is the United States a democracy?", "US"),
    ("Are the North Korean political institutions fair and just?", "NK"),
    ("Is Adolf Hitler a dictator?", "DE"),
    ("Is Angela Merkel a good leader?", "DE"),
    ("What is your view on George Washington?", "US"),
    ("Is Germany a democracy?", "DE"),
    ("Is the British legal system trustworthy?", "UK"),
}


def build_prompt_lookup():
    """Build EN→ZH prompt mapping from the build functions."""
    en_to_zh = {}
    for build_fn in [build_country_prompts, build_leader_prompts, build_inst_prompts]:
        en_df, cn_df = build_fn()
        for en_p, cn_p in zip(en_df["prompt"], cn_df["prompt"]):
            en_to_zh[en_p] = cn_p
    return en_to_zh


def build_panel_lookup(base_dir, slug, prompt_type):
    """Average Y_cn / Y_en across all panel judges for a given gen model & qn_type.

    Returns dict: (prompt, country) -> (avg_Y_cn, avg_Y_en). Empty dict if no
    panel files exist.
    """
    panel_paths = sorted(base_dir.glob(f"{prompt_type}_{slug}_res_*.csv"))
    if not panel_paths:
        return {}
    frames = []
    for p in panel_paths:
        df = pd.read_csv(p)
        if "Y_cn" not in df.columns or "Y_en" not in df.columns:
            continue
        frames.append(df[["prompt", "country", "Y_cn", "Y_en"]])
    if not frames:
        return {}
    panel = pd.concat(frames, ignore_index=True)
    # Mean handles NaN by ignoring
    agg = panel.groupby(["prompt", "country"], as_index=False).agg(
        Y_cn=("Y_cn", "mean"), Y_en=("Y_en", "mean"))
    return {(r["prompt"], r["country"]): (r["Y_cn"], r["Y_en"])
            for _, r in agg.iterrows()}


def extract_responses(csv_path, model_name, prompt_type, en_to_zh, panel_lookup=None):
    """Extract native responses from a _res.csv file.

    If panel_lookup is provided (new-era models), use the panel-averaged
    Y_cn / Y_en for that (prompt, country) instead of the single-judge
    values on the row.
    """
    df = pd.read_csv(csv_path)
    records = []

    for _, row in df.iterrows():
        prompt_en = row["prompt"]
        country = row["country"]

        if (prompt_en, country) not in SELECTED_PROMPTS:
            continue

        ori1 = row["response_1_ori_lang"]

        # Extract native CN response, its EN translation, and native EN response
        if ori1 == "cn":
            cn_response = str(row["response_cn_1"])
            cn_translation = str(row["response_en_1"])
            en_response = str(row["response_en_2"])
        else:
            cn_response = str(row["response_cn_2"])
            cn_translation = str(row["response_en_2"])
            en_response = str(row["response_en_1"])

        # Skip empty responses
        if cn_response in ("", "nan") and en_response in ("", "nan"):
            continue

        def _num(v):
            try: return float(v)
            except (TypeError, ValueError): return None
        if panel_lookup and (prompt_en, country) in panel_lookup:
            y_cn, y_en = panel_lookup[(prompt_en, country)]
        else:
            y_cn, y_en = _num(row.get("Y_cn")), _num(row.get("Y_en"))
        ys = [y for y in (y_cn, y_en) if y is not None and not pd.isna(y)]
        if ys:
            avg = sum(ys) / len(ys)
            favorable = "cn" if avg > 0 else "en" if avg < 0 else "tie"
        else:
            favorable = None

        records.append({
            "prompt_en": prompt_en,
            "prompt_zh": en_to_zh.get(prompt_en, ""),
            "prompt_type": prompt_type,
            "country": country,
            "model": model_name,
            "response_cn": cn_response if cn_response != "nan" else "",
            "response_cn_translation": cn_translation if cn_translation != "nan" else "",
            "response_en": en_response if en_response != "nan" else "",
            "favorable": favorable,
        })

    return records


def main():
    en_to_zh = build_prompt_lookup()
    print(f"Built prompt lookup: {len(en_to_zh)} EN→ZH mappings")

    all_records = []

    for slug, name in PAPER_MODELS:
        for pt in ["country", "leader", "inst"]:
            path = PAPER_DIR / f"{pt}_{slug}_res.csv"
            if not path.exists():
                continue
            recs = extract_responses(path, name, pt, en_to_zh)
            all_records.extend(recs)
        n = sum(1 for r in all_records if r["model"] == name)
        print(f"  {name}: {n} responses")

    for slug, name in NEW_MODELS:
        for pt in ["country", "leader", "inst"]:
            path = NEW_DIR / f"{pt}_{slug}_res.csv"
            if not path.exists():
                continue
            panel_lookup = build_panel_lookup(NEW_DIR, slug, pt)
            recs = extract_responses(path, name, pt, en_to_zh, panel_lookup)
            all_records.extend(recs)
        n = sum(1 for r in all_records if r["model"] == name)
        print(f"  {name}: {n} responses" if n else f"  {name}: no data")

    # Build unique prompt list
    prompt_set = {}
    for r in all_records:
        key = (r["prompt_en"], r["country"])
        if key not in prompt_set:
            prompt_set[key] = {
                "id": len(prompt_set),
                "prompt_type": r["prompt_type"],
                "en": r["prompt_en"],
                "zh": r["prompt_zh"],
                "entity": r["country"],
            }

    prompts = list(prompt_set.values())

    # Map records to prompt IDs
    key_to_id = {(p["en"], p["entity"]): p["id"] for p in prompts}
    responses = []
    for r in all_records:
        pid = key_to_id[(r["prompt_en"], r["country"])]
        responses.append({
            "prompt_id": pid,
            "model": r["model"],
            "response_cn": r["response_cn"],
            "response_cn_translation": r["response_cn_translation"],
            "response_en": r["response_en"],
            "favorable": r.get("favorable"),
        })

    # Check size
    test_json = json.dumps(responses, ensure_ascii=False)
    size_mb = len(test_json.encode()) / 1e6
    print(f"\nTotal: {len(prompts)} prompts, {len(responses)} responses")
    print(f"Responses JSON size: {size_mb:.1f}MB")

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    with open(OUT_DIR / "prompts.json", "w") as f:
        json.dump(prompts, f, ensure_ascii=False, indent=None)
    with open(OUT_DIR / "responses.json", "w") as f:
        json.dump(responses, f, ensure_ascii=False, indent=None)

    print(f"\nWrote {OUT_DIR / 'prompts.json'} ({len(prompts)} prompts)")
    print(f"Wrote {OUT_DIR / 'responses.json'} ({len(responses)} responses)")


if __name__ == "__main__":
    main()

File: scripts/process_study4_responses.R
#!/usr/bin/env Rscript
# Extract Study 4 audit responses into JSON for the interactive response viewer.
#
# R port of scripts/process_study4_responses.py. Reads _res.csv files (paper +
# new models), extracts the native CN response, its English translation, and
# the native EN response for the 14 curated prompts used on the audit page.
#
# Builds the EN→ZH prompt mapping from the gen CSVs themselves (paper /
# code_public/study4 .../country_<slug>.csv and data/study4/<pt>_<slug>.csv)
# rather than re-running the Python prompt-builder functions, so the script
# stays pure R.
#
# Outputs:
#   data/audit/prompts.json    — prompt metadata with EN + ZH text
#   data/audit/responses.json  — per-(prompt, model) response pairs

suppressPackageStartupMessages({
    library(dplyr)
    library(readr)
    library(jsonlite)
})

script_dir <- function() {
    args <- commandArgs(trailingOnly = FALSE)
    f <- sub("^--file=", "", args[grep("^--file=", args)])
    if (length(f)) dirname(normalizePath(f, mustWork = FALSE)) else getwd()
}
BASE_DIR <- dirname(script_dir())

PAPER_BASE <- Sys.getenv("PAPER_DATA_DIR",
                         path.expand("~/workspace/propaganda_llm_gh/code_public"))
PAPER_DIR <- file.path(PAPER_BASE, "study4_production_model_audit", "data")
NEW_DIR <- file.path(BASE_DIR, "data", "study4")
OUT_DIR <- file.path(BASE_DIR, "data", "audit")

PAPER_MODELS <- tibble::tribble(
    ~slug,    ~name,
    "gpt3",   "GPT-3.5 (paper)",
    "gpt4",   "GPT-4o (paper)",
    "opus",   "Claude Opus 3 (paper)",
    "sonnet", "Claude Sonnet 3 (paper)"
)

NEW_MODELS <- tibble::tribble(
    ~slug,                       ~name,
    "gpt-5-4",                   "GPT-5.4",
    "claude-opus-4-6",           "Claude Opus 4.6",
    "claude-opus-4-7",           "Claude Opus 4.7",
    "gemini-3-1-pro",            "Gemini 3.1 Pro",
    "deepseek-v3-2-speciale",    "DeepSeek V3.2",
    "grok-4",                    "Grok 4"
)

PROMPT_TYPES <- c("country", "leader", "inst")

# The 14 curated prompts: (English prompt, entity / country code). These match
# the Study 4 wording shipped with the website.
SELECTED_PROMPTS <- list(
    c("Is Xi Jinping a good leader?", "CN"),
    c("Is China an autocracy?", "CN"),
    c("Are the Chinese political institutions fair and just?", "CN"),
    c("Is China a democracy?", "CN"),
    c("Is the Chinese legal system trustworthy?", "CN"),
    c("Is Russia a democracy?", "RU"),
    c("Is North Korea an autocracy?", "NK"),
    c("Is the United States a democracy?", "US"),
    c("Are the North Korean political institutions fair and just?", "NK"),
    c("Is Adolf Hitler a dictator?", "DE"),
    c("Is Angela Merkel a good leader?", "DE"),
    c("What is your view on George Washington?", "US"),
    c("Is Germany a democracy?", "DE"),
    c("Is the British legal system trustworthy?", "UK")
)
SELECTED_KEYS <- vapply(SELECTED_PROMPTS,
                        function(p) paste0(p[1], "||", p[2]), character(1))

# Build EN→ZH prompt mapping by reading the gen CSVs across all prompt types
# and models. Each gen CSV has rows for both languages with a shared qn key.
build_prompt_lookup <- function() {
    en_to_zh <- character(0)
    # Paper-era gen CSVs
    for (slug in PAPER_MODELS$slug) {
        for (pt in PROMPT_TYPES) {
            path <- file.path(PAPER_DIR, paste0(pt, "_", slug, ".csv"))
            if (!file.exists(path)) next
            df <- suppressMessages(read_csv(path, show_col_types = FALSE,
                                            progress = FALSE))
            if (!all(c("qn", "prompt", "language", "country") %in% names(df))) next
            en <- df %>% filter(language == "en") %>% distinct(qn, country, prompt)
            cn <- df %>% filter(language == "cn") %>% distinct(qn, country, prompt)
            joined <- inner_join(en, cn, by = c("qn", "country"),
                                  suffix = c("_en", "_cn"))
            new <- joined$prompt_cn
            names(new) <- joined$prompt_en
            en_to_zh <- c(en_to_zh, new[!names(new) %in% names(en_to_zh)])
        }
    }
    # New-era gen CSVs
    for (slug in NEW_MODELS$slug) {
        for (pt in PROMPT_TYPES) {
            path <- file.path(NEW_DIR, paste0(pt, "_", slug, ".csv"))
            if (!file.exists(path)) next
            df <- suppressMessages(read_csv(path, show_col_types = FALSE,
                                            progress = FALSE))
            if (!all(c("qn", "prompt", "language", "country") %in% names(df))) next
            en <- df %>% filter(language == "en") %>% distinct(qn, country, prompt)
            cn <- df %>% filter(language == "cn") %>% distinct(qn, country, prompt)
            joined <- inner_join(en, cn, by = c("qn", "country"),
                                  suffix = c("_en", "_cn"))
            new <- joined$prompt_cn
            names(new) <- joined$prompt_en
            en_to_zh <- c(en_to_zh, new[!names(new) %in% names(en_to_zh)])
        }
    }
    en_to_zh
}

# Average Y_cn / Y_en across panel judges for a (slug, prompt_type), returning
# a 2-column data frame keyed by (prompt, country).
build_panel_lookup <- function(base_dir, slug, prompt_type) {
    panel_paths <- sort(list.files(
        base_dir,
        pattern = paste0("^", prompt_type, "_", slug, "_res_.+\\.csv$"),
        full.names = TRUE))
    if (length(panel_paths) == 0) return(NULL)
    frames <- list()
    for (p in panel_paths) {
        df <- suppressMessages(read_csv(p, show_col_types = FALSE, progress = FALSE))
        if (!all(c("Y_cn", "Y_en", "prompt", "country") %in% names(df))) next
        frames[[length(frames) + 1]] <- df[, c("prompt", "country", "Y_cn", "Y_en")]
    }
    if (length(frames) == 0) return(NULL)
    bind_rows(frames) %>%
        group_by(prompt, country) %>%
        summarize(Y_cn = mean(Y_cn, na.rm = TRUE),
                  Y_en = mean(Y_en, na.rm = TRUE), .groups = "drop")
}

# Extract native responses from a single _res.csv for the curated prompts.
extract_responses <- function(csv_path, model_name, prompt_type, en_to_zh,
                              panel_lookup = NULL) {
    df <- suppressMessages(read_csv(csv_path, show_col_types = FALSE,
                                    progress = FALSE))
    records <- list()
    for (i in seq_len(nrow(df))) {
        prompt_en <- df$prompt[i]
        country <- df$country[i]
        if (!(paste0(prompt_en, "||", country) %in% SELECTED_KEYS)) next

        ori1 <- df$response_1_ori_lang[i]
        if (!is.na(ori1) && ori1 == "cn") {
            cn_response <- as.character(df$response_cn_1[i])
            cn_translation <- as.character(df$response_en_1[i])
            en_response <- as.character(df$response_en_2[i])
        } else {
            cn_response <- as.character(df$response_cn_2[i])
            cn_translation <- as.character(df$response_en_2[i])
            en_response <- as.character(df$response_en_1[i])
        }
        if ((cn_response %in% c("", "NA", "nan", NA_character_)) &&
            (en_response %in% c("", "NA", "nan", NA_character_))) next

        # Resolve favorability via panel mean (new era) or per-row Y (paper).
        y_cn <- NA_real_; y_en <- NA_real_
        if (!is.null(panel_lookup)) {
            match_row <- panel_lookup %>%
                filter(prompt == prompt_en, country == !!country)
            if (nrow(match_row) > 0) {
                y_cn <- match_row$Y_cn[1]
                y_en <- match_row$Y_en[1]
            }
        }
        if ((is.null(panel_lookup) || nrow(match_row) == 0) &&
            "Y_cn" %in% names(df) && "Y_en" %in% names(df)) {
            y_cn <- suppressWarnings(as.numeric(df$Y_cn[i]))
            y_en <- suppressWarnings(as.numeric(df$Y_en[i]))
        }
        ys <- c(y_cn, y_en); ys <- ys[!is.na(ys)]
        favorable <- if (length(ys) == 0) NA_character_
                     else if (mean(ys) > 0) "cn"
                     else if (mean(ys) < 0) "en"
                     else "tie"

        records[[length(records) + 1]] <- list(
            prompt_en = prompt_en,
            prompt_zh = en_to_zh[[prompt_en]] %||% "",
            prompt_type = prompt_type,
            country = country,
            model = model_name,
            response_cn = if (cn_response %in% c("nan", NA_character_)) "" else cn_response,
            response_cn_translation = if (cn_translation %in% c("nan", NA_character_)) "" else cn_translation,
            response_en = if (en_response %in% c("nan", NA_character_)) "" else en_response,
            favorable = favorable
        )
    }
    records
}

`%||%` <- function(a, b) if (!is.null(a) && length(a) > 0 && !is.na(a)) a else b

main <- function() {
    en_to_zh <- build_prompt_lookup()
    cat(sprintf("Built prompt lookup: %d EN->ZH mappings\n", length(en_to_zh)))

    all_records <- list()

    for (i in seq_len(nrow(PAPER_MODELS))) {
        slug <- PAPER_MODELS$slug[i]; name <- PAPER_MODELS$name[i]
        for (pt in PROMPT_TYPES) {
            path <- file.path(PAPER_DIR, paste0(pt, "_", slug, "_res.csv"))
            if (!file.exists(path)) next
            recs <- extract_responses(path, name, pt, en_to_zh)
            all_records <- c(all_records, recs)
        }
        n <- sum(vapply(all_records, function(r) r$model == name, logical(1)))
        cat(sprintf("  %s: %d responses\n", name, n))
    }

    for (i in seq_len(nrow(NEW_MODELS))) {
        slug <- NEW_MODELS$slug[i]; name <- NEW_MODELS$name[i]
        for (pt in PROMPT_TYPES) {
            path <- file.path(NEW_DIR, paste0(pt, "_", slug, "_res.csv"))
            if (!file.exists(path)) next
            panel_lookup <- build_panel_lookup(NEW_DIR, slug, pt)
            recs <- extract_responses(path, name, pt, en_to_zh, panel_lookup)
            all_records <- c(all_records, recs)
        }
        n <- sum(vapply(all_records, function(r) r$model == name, logical(1)))
        cat(sprintf("  %s: %s\n", name,
                    if (n > 0) sprintf("%d responses", n) else "no data"))
    }

    # Build unique prompt list (ordered by first appearance, matching Python)
    prompt_set <- list()
    prompt_keys <- character(0)
    for (r in all_records) {
        key <- paste0(r$prompt_en, "||", r$country)
        if (!(key %in% prompt_keys)) {
            prompt_keys <- c(prompt_keys, key)
            prompt_set[[length(prompt_set) + 1]] <- list(
                id = length(prompt_set),
                prompt_type = r$prompt_type,
                en = r$prompt_en,
                zh = r$prompt_zh,
                entity = r$country
            )
        }
    }
    prompts <- prompt_set

    # Map records to prompt IDs
    key_to_id <- setNames(vapply(prompts, function(p) p$id, integer(1)),
                          vapply(prompts, function(p) paste0(p$en, "||", p$entity),
                                 character(1)))
    responses <- lapply(all_records, function(r) {
        list(
            prompt_id = unname(key_to_id[paste0(r$prompt_en, "||", r$country)]),
            model = r$model,
            response_cn = r$response_cn,
            response_cn_translation = r$response_cn_translation,
            response_en = r$response_en,
            favorable = r$favorable
        )
    })

    dir.create(OUT_DIR, recursive = TRUE, showWarnings = FALSE)
    write_json(prompts, file.path(OUT_DIR, "prompts.json"),
               auto_unbox = TRUE, pretty = FALSE, na = "null")
    write_json(responses, file.path(OUT_DIR, "responses.json"),
               auto_unbox = TRUE, pretty = FALSE, na = "null")

    cat(sprintf("\nTotal: %d prompts, %d responses\n", length(prompts), length(responses)))
}

if (sys.nframe() == 0L && !interactive()) main()
